### PR TITLE
add MQTT client

### PR DESCRIPTION
**Pull request by arpiecodes**
This pull is my second attempt to add MQTT support (see https://github.com/serkri/SmartEVSE-3/pull/69 for first attempt). It was kept to the minimum to enable MQTT to function properly this time, and only contains minor refactors where it made sense to do so.

The pull request seems to contain a lot of added code, but this is due to the addition of a lightweight MQTT client library (I simply included the whole Git repo contents as it was not possible to directly use it through pio dependencies). The standard Arduino MQTTPubSubClient library would require a Ethernet/WiFiClient, and it seems we are not using that in this project with mixed Arduino/ESP IDF framework code. If someone figures out a better way to implement it properly using standard libraries, contributions to the code are very welcome. I also edited the library to support username/password authentication.

Code has been tested on my SmartEVSE for some time already albeit in another form. "It works on my machine" applies, YMMV, some testing would probably be wise.

The MQTT support can easily be disabled again by defining `MQTT_DISABLED`.

PS: Swagger will need an update after this is merged to document all the new MQTT setting params. I am willing to take a stab at that when the code is accepted and merged.

### CODE REVIEW

**Comment by dingo35** [comment]
@arpiecodes First of all, love your coding style, looks like a really clean job!

I understand your struggle with the libraries, been there, done that. I understand you didn't want to include the library, but you had no choice. What I did not like is that you had to add set_credentials, so now we have an unmaintained, modified library.
Still, I have no problem to do that as long as it is our last resort. 

I did not understand the ethernet/wifi client remark (since we use that wifi.h library), so I fiddled around a bit with the MQTTPubSubClient library; in the zip file you find a .diff that shows how I think the MQTTPubSubClient lib could be used; the code compiles but is otherwise UNTESTED since I currently have no MQTT server running; it will not work since the callback code has to be moved and I didn't want to make the diff unreadable.

Please let me know what you think, would moving towards that lib be feasible? Pro's, cons?

[mqtt_pubsub.zip](https://github.com/serkri/SmartEVSE-3/files/11879549/mqtt_pubsub.zip)

**Comment by arpiecodes** [comment]
> 

Nice! I'll check this out and confirm whether it works in a 'MQTT production set-up'. It would definitely make things a lot more 'clean' and we incur less technical debt.

I think you found a more generic version of the same library that supports more platforms. The one I found was the original Arduino one. Any ways; it looks good. Let me report back soon.

**Comment by arpiecodes** [comment]
Also, good to note; this pull request is missing the 'EVState' specific Publish and Set subjects since these variables are introduced in https://github.com/serkri/SmartEVSE-3/pull/160 respectively.

**Comment by dingo35** [comment]
Ok thanks for the bug discovery!!

Could you upload your code (just zip total SmartEVSE dir or a diff to current master) so I can take a look?

**Comment by dingo35** [comment]
... you might also try this library, which is derived from the one above, so it shouldn't need much changes in your code:
https://github.com/khoih-prog/MQTTPubSubClient_Generic

It is an archived library, but this khoih guy makes good stuff.....

**Comment by arpiecodes** [comment]
@dingo35 I found another library (https://github.com/256dpi/arduino-mqtt) that seems to be actively supported with a healthy number of stars. It's also easily included into the project with pio. So I went with that. It seems to work nice now!

Also performed some minor refactors, removed some stuff I was not sure of to include as of yet. Also tested some edge cases, all seems to go well. As far as I'm concerned, it's good to go.

**Comment by dingo35** [comment]
@arpiecodes Ok I'm putting up a test environment; first mistake was to use the public test.mosquitto.org server, hundreds of devices appeared in my HA; so I setup a local mqtt server, it receives topics from SmartEVSE, I can subscribe to topics in HA, but how do I get the autodiscovery activated? It must be some setting in the mqtt server, but I cannot find it....

**Comment by dingo35** [comment]
Ok great! Do you mind if I change ifndef MQTT_DISABLE  to ifdef MQTT  ? I find double negatives hard to read.... IMHO MQTT should be enabled by default, because it adds responsiveness to HA while not costing much in resources....

**Comment by arpiecodes** [comment]
I have some more changes coming up actually, which I will push later today. I think it's worth the wait before we merge.

**Comment by dingo35** [comment]
O ok, Im currently doing some minor layout / editing changes directly in github, that Ok or you want me to await your push?

**Comment by arpiecodes** [comment]
> O ok, Im currently doing some minor layout / editing changes directly in github, that Ok or you want me to await your push?

If you could add them as review comments in-line, I will be sure to handle them as I push later. 👍

**Comment by dingo35** [comment]
Ok will do!

File: SmartEVSE-3/include/evse.h [hunk 1]
```diff
@@ -30,6 +30,10 @@
 #define DBG 0  //comment or set to 0 for production release, 0 = no debug 1 = debug over telnet, 2 = debug over usb AND telnet
 #endif
 
+#ifndef MQTT_DISABLED
+// #define MQTT_DISABLED 1                                                      // Uncomment to disable MQTT support in code
```
Comment (dingo35) [inline SmartEVSE-3/include/evse.h:34]: Please: 1,$s/ifndef MQTT_DISABLED/ifdef MQTT/g

and change //#define MQTT_DISABLED to:
#define MQTT 1 //comment or set to 0 to disable MQTT

...so it is more in line with the other compile options.

File: SmartEVSE-3/data/index.html [hunk 2]
```diff
@@ -244,6 +254,14 @@
...
             }
 
+            if (data.mqtt && !mqttEditMode) {
+                $('#mqtt_host').val(data.mqtt.broker_host);
+                $('#mqtt_port').val(data.mqtt.broker_port);
+                $('#mqtt_username').val(data.mqtt.username);
+                $('#mqtt_password').val(data.mqtt.password);
+                $('#mqtt_topic_prefix').val(data.mqtt.topic_prefix);
+            }
+
```
Comment (dingo35) [inline SmartEVSE-3/data/index.html:264]: Please remove space on this empty line

File: SmartEVSE-3/src/evse.cpp [hunk 3]
```diff
@@ -584,10 +629,10 @@ void setMode(uint8_t NewMode) {
         if ((Mode != MODE_SOLAR && NewMode == MODE_SOLAR) || (Mode == MODE_SOLAR && NewMode != MODE_SOLAR)) {
             //we are switching from non-solar to solar
             //since we EnableC2 == SOLAR_OFF C2 is turned On now, and should be turned off
-            setAccess(0);                                                       //switch to OFF
+            setAccess(false);                                                       //switch to OFF
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:632]: Unnecessary change, distracts from your MQTT code. Please undo.

File: SmartEVSE-3/src/evse.cpp [hunk 4]
```diff
@@ -584,10 +629,10 @@ void setMode(uint8_t NewMode) {
...
             if (LoadBl == 1) ModbusWriteSingleRequest(BROADCAST_ADR, 0x0003, NewMode);
             Mode = NewMode;
-            setAccess(1);
+            setAccess(true);
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:635]: Unnecessary change, distracts from your MQTT code. Please undo.

File: SmartEVSE-3/src/evse.cpp [hunk 5]
```diff
@@ -2110,6 +2164,210 @@ uint8_t PollEVNode = NR_EVSES;
 
 }
 
+#ifndef MQTT_DISABLED
+void mqtt_receive_callback(const String &topic, const String &payload)
+{
+    if (topic == MQTTprefix + "/Set/Mode")
+    {
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:2171]: General layout remark: please keep opening { on the same line as the if statement....
So: if (topic == MQTTprefix) {
            //something 
      }

this saves (almost) empty lines, which fits more lines of code on your screen.....

File: SmartEVSE-3/src/evse.cpp [hunk 6]
```diff
@@ -2110,6 +2164,210 @@ uint8_t PollEVNode = NR_EVSES;
...
+
+    // Make sure MQTT updates directly to prevent debounces
+    lastMqttUpdate = 10;
+}
+
+void SetupMQTTClient()
+{
+    // Disconnect existing connection if connected
+    if (MQTTclient.connected())
+    {
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:2298]: In these "simple", unambiguous if statements, please leave out the { and }, saves lines and make code more readable.
So:
if (MQTTclient.connected())
      MQTTclient.disconnect();

File: SmartEVSE-3/src/evse.cpp [hunk 7]
```diff
@@ -2110,6 +2164,210 @@ uint8_t PollEVNode = NR_EVSES;
...
+        MQTTclient.disconnect();
+    }
+
+    // No need to attempt connections if we aren't configured
+    if (MQTTHost == "")
+    {
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:2304]: see remark above...

File: SmartEVSE-3/src/evse.cpp [hunk 8]
```diff
@@ -2110,6 +2164,210 @@ uint8_t PollEVNode = NR_EVSES;
...
+    // Setup and connect MQTT client instance
+    MQTTclient.setHost(MQTTHost.c_str(), MQTTPort);
+
+    if (MQTTuser != "" && MQTTpassword != "")
+    {
+        MQTTclient.connect(APhostname.c_str(), MQTTuser.c_str(), MQTTpassword.c_str());
+    } else {
+        MQTTclient.connect(APhostname.c_str());
+    }
+
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:2317]: see above

File: SmartEVSE-3/src/evse.cpp [hunk 9]
```diff
@@ -3383,34 +3672,21 @@ void StartwebServer(void) {
                     ToModemWaitStateTimer = 0;
                     ToModemDoneStateTimer = 0;
                     LeaveModemDoneStateTimer = 0;
-                    setAccess(0);
+                    setAccess(false);
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:3675]: Unnecessary change, distracts from your MQTT code. Please undo.

File: SmartEVSE-3/src/evse.cpp [hunk 10]
```diff
@@ -3383,34 +3672,21 @@ void StartwebServer(void) {
...
                     break;
                 case 2:
-                    if (MainsMeter) {
-                        OverrideCurrent = 0;
-                        setAccess(!DelayedStartTime.epoch2);
-                        setMode(MODE_SOLAR);
-                        break;
-                    }
+                    setMode(MODE_SOLAR);
+                    break;
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:3682]: These two lines seem incorrect to me: if you have no mainsmeter installed, you can not run in Solar mode!

File: SmartEVSE-3/src/evse.cpp [hunk 11]
```diff
@@ -3383,34 +3672,21 @@ void StartwebServer(void) {
...
                 case 3:
-                    if (MainsMeter) {
-                        setAccess(!DelayedStartTime.epoch2);
-                        setMode(MODE_SMART);
-                        break;
-                    }
+                    setMode(MODE_SMART);
+                    break;
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:3685]: These two lines seem incorrect to me: if you have no mainsmeter installed, you can not run in Smart mode!

File: SmartEVSE-3/src/evse.cpp [hunk 12]
```diff
@@ -3997,15 +4324,15 @@ void loop() {
         DelayedStartTime.diff = DelayedStartTime.epoch2 - (mktime(localtime(&now)) - EPOCH2_OFFSET);
         if (DelayedStartTime.diff > 0) {
             if (Access_bit != 0 && (DelayedStopTime.epoch2 == 0 || DelayedStopTime.epoch2 > DelayedStartTime.epoch2))
-                setAccess(0);                         //switch to OFF, we are Delayed Charging
+                setAccess(false);                         //switch to OFF, we are Delayed Charging
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:4327]: Unnecessary change, distracts from your MQTT code. Please undo.

File: SmartEVSE-3/src/evse.cpp [hunk 13]
```diff
@@ -3997,15 +4324,15 @@ void loop() {
...
         }
         else {
             //starttime is in the past so we are NOT Delayed Charging, or we are Delayed Charging but the starttime has passed!
             if (DelayedRepeat == 1)
                 DelayedStartTime.epoch2 += 24 * 3600;                           //add 24 hours so we now have a new starttime
             else
                 DelayedStartTime.epoch2 = DELAYEDSTARTTIME;
-            setAccess(1);
+            setAccess(true);
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:4335]: Unnecessary change, distracts from your MQTT code. Please undo.

File: SmartEVSE-3/src/evse.cpp [hunk 14]
```diff
@@ -4019,7 +4346,7 @@ void loop() {
                 DelayedStopTime.epoch2 += 24 * 3600;                        //add 24 hours so we now have a new starttime
             else
                 DelayedStopTime.epoch2 = DELAYEDSTOPTIME;
-            setAccess(0);                         //switch to OFF
+            setAccess(false);                         //switch to OFF
```
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:4349]: Unnecessary change, distracts from your MQTT code. Please undo.

**Review by dingo35 — CHANGES_REQUESTED** [review]
Just minor remarks, love your coding style!

File: SmartEVSE-3/src/evse.cpp (hunk 3 above)
Comment (arpiecodes) [reply SmartEVSE-3/src/evse.cpp:632]: I agree, seems to have been a remnant of me trying to debug why setAccess worked unreliably. I'll put them back how they were.

File: SmartEVSE-3/src/evse.cpp (hunk 5 above)
Comment (arpiecodes) [reply SmartEVSE-3/src/evse.cpp:2171]: Normally I'd agree, but while working on this project it really helped to have the extra tidy-ness it brings while trying to understand the somehow somewhat convoluted logic. I suppose it's a matter of opinion, so I'll make it in line with the rest of the code.

File: SmartEVSE-3/src/evse.cpp (hunk 6 above)
Comment (arpiecodes) [reply SmartEVSE-3/src/evse.cpp:2298]: Same as above actually; but I'll make it so.

File: SmartEVSE-3/src/evse.cpp (hunk 10 above)
Comment (arpiecodes) [reply SmartEVSE-3/src/evse.cpp:3682]: That check is already in `setMode`, and since MQTT does not really reply implicitly when handling the set less code seemed a better outcome to me.

File: SmartEVSE-3/src/evse.cpp (hunk 11 above)
Comment (arpiecodes) [reply SmartEVSE-3/src/evse.cpp:3685]: Same as with SOLAR.

File: SmartEVSE-3/src/evse.cpp (hunk 10 above)
Comment (dingo35) [reply SmartEVSE-3/src/evse.cpp:3682]: Correct, thanks!! 
Like Elon said, Id like to pay my dev's for every line of code they write, but I'd pay them double for every line they remove!

**Review by dingo35 — APPROVED** [review]

**Comment by dingo35** [comment]
@arpiecodes Just pushed the discovery-code for the sensors ....

**Comment by dingo35** [comment]
@arpiecodes any comment on PR #172 ? I'm considering committing it, but also to remove EVChargeCurrent topic since it is a ducplicate of ChargeCurrent....

What do you think?
